### PR TITLE
buckingham palace invite form

### PR DESCRIPTION
**Pull request by dreamfall**

### CODE REVIEW

File: app/controllers/palace_invites_controller.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  def load_invite
+    palace_invite = PalaceInvite.find_by_token(params[:id])
+    @invite_form = PalaceInviteForm.new(palace_invite)
+  end
+
+  def palace_invite_attributes
+    params.require(:palace_invite).permit(
+      palace_attendees_attributes:
+        [:title,
+        :first_name,
```
Comment (matthewford) [inline app/controllers/palace_invites_controller.rb]: Align the elements of an array literal if they span more than one line.

File: app/controllers/palace_invites_controller.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        :last_name,
```
Comment (matthewford) [inline app/controllers/palace_invites_controller.rb]: Align the elements of an array literal if they span more than one line.

File: app/controllers/palace_invites_controller.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        :job_name,
```
Comment (matthewford) [inline app/controllers/palace_invites_controller.rb]: Align the elements of an array literal if they span more than one line.

File: app/controllers/palace_invites_controller.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        :post_nominals,
```
Comment (matthewford) [inline app/controllers/palace_invites_controller.rb]: Align the elements of an array literal if they span more than one line.

File: app/controllers/palace_invites_controller.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        :address_1,
```
Comment (matthewford) [inline app/controllers/palace_invites_controller.rb]: Align the elements of an array literal if they span more than one line.

File: app/controllers/palace_invites_controller.rb [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        :address_2,
```
Comment (matthewford) [inline app/controllers/palace_invites_controller.rb]: Align the elements of an array literal if they span more than one line.

File: app/controllers/palace_invites_controller.rb [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        :address_3,
```
Comment (matthewford) [inline app/controllers/palace_invites_controller.rb]: Align the elements of an array literal if they span more than one line.

File: app/controllers/palace_invites_controller.rb [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        :address_4,
```
Comment (matthewford) [inline app/controllers/palace_invites_controller.rb]: Align the elements of an array literal if they span more than one line.

File: app/controllers/palace_invites_controller.rb [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        :postcode,
```
Comment (matthewford) [inline app/controllers/palace_invites_controller.rb]: Align the elements of an array literal if they span more than one line.

File: app/controllers/palace_invites_controller.rb [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+        :phone_number,
```
Comment (matthewford) [inline app/controllers/palace_invites_controller.rb]: Align the elements of an array literal if they span more than one line.

File: app/controllers/palace_invites_controller.rb [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+        :product_description,
```
Comment (matthewford) [inline app/controllers/palace_invites_controller.rb]: Align the elements of an array literal if they span more than one line.

File: app/controllers/palace_invites_controller.rb [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+        :additional_info,
```
Comment (matthewford) [inline app/controllers/palace_invites_controller.rb]: Align the elements of an array literal if they span more than one line.

File: app/controllers/palace_invites_controller.rb [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+        :id,
```
Comment (matthewford) [inline app/controllers/palace_invites_controller.rb]: Align the elements of an array literal if they span more than one line.

File: app/controllers/palace_invites_controller.rb [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+        :_remove]
```
Comment (matthewford) [inline app/controllers/palace_invites_controller.rb]: Align the elements of an array literal if they span more than one line.

File: app/forms/palace_invite_form.rb [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+    false
+  end
+
+  def palace_attendees_attributes=(attrs = {})
+    attrs.each do |(_, attendee_attributes)|
+      attendee = if attendee_attributes["id"]
+        palace_attendees.detect { |a| a.id.to_s == attendee_attributes["id"] }
+      else
+        build_palace_attendee
+      end
```
Comment (matthewford) [inline app/forms/palace_invite_form.rb]: `end` at 44, 6 is not aligned with `if` at 40, 17

File: app/forms/palace_invite_form.rb [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+    @attendees ||= []
+    @attendees << invite.palace_attendees.build
+    @attendees.last
+  end
+
+  def persisted?
+    false
+  end
+
+  def palace_attendees_attributes=(attrs = {})
```
Comment (matthewford) [inline app/forms/palace_invite_form.rb]: Assignment Branch Condition size for palace_attendees_attributes= is too high. [15.33/15]

File: app/forms/palace_invite_form.rb [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+    attrs.each do |(_, attendee_attributes)|
+      attendee = if attendee_attributes["id"]
+        palace_attendees.detect { |a| a.id.to_s == attendee_attributes["id"] }
+      else
```
Comment (matthewford) [inline app/forms/palace_invite_form.rb]: Align `else` with `if`.

File: app/forms/palace_invite_form.rb [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+  end
+
+  def persisted?
+    false
+  end
+
+  def palace_attendees_attributes=(attrs = {})
+    attrs.each do |(_, attendee_attributes)|
+      attendee = if attendee_attributes["id"]
+        palace_attendees.detect { |a| a.id.to_s == attendee_attributes["id"] }
```
Comment (matthewford) [inline app/forms/palace_invite_form.rb]: Use 2 (not -9) spaces for indentation.

File: app/mailers/users/buckingham_palace_invite_mailer.rb [hunk 19]
```diff
@@ -0,0 +1,9 @@
+class Users::BuckinghamPalaceInviteMailer < ApplicationMailer
+  def invite(invite_id)
+    invite = PalaceInvite.find(invite_id)
+    @token = invite.token
+
+    @subject = "[Queen's Awards] #{invite.form_answer.decorate.award_type} request to fill in attendee details"
```
Comment (matthewford) [inline app/mailers/users/buckingham_palace_invite_mailer.rb]: Line is too long. [111/100]

File: app/models/form_answer.rb [hunk 20]
```diff
@@ -90,6 +90,7 @@ def secondary
     scope :for_award_type, -> (award_type) { where award_type: award_type }
     scope :for_year, -> (year) { where award_year: year }
     scope :shortlisted_with_no_certificate, -> { where("1 = 0") }
+    scope :winners, -> { where("1 = 0") }
```
Comment (matthewford) [inline app/models/form_answer.rb]: Inconsistent indentation detected.

File: app/renderers/mail_renderer.rb [hunk 21]
```diff
@@ -19,9 +19,9 @@ def default_url_options
   def shortlisted_audit_certificate_reminder
     assigns = {}
 
-    assigns[:form_owner] = User.new(first_name: 'Jon', last_name: 'Doe').decorate
-    assigns[:recipient] = User.new(first_name: 'Jane', last_name: 'Doe').decorate
-    assigns[:form_answer] = FormAnswer.new(id: 0, user: assigns[:user], award_type: 'promotion').decorate
+    assigns[:form_owner] = dummy_user("Jon", "Doe")
+    assigns[:recipient] = dummy_user("Jane", "Doe")
+    assigns[:form_answer] = FormAnswer.new(id: 0, user: assigns[:user], award_type: "promotion").decorate
```
Comment (matthewford) [inline app/renderers/mail_renderer.rb]: Line is too long. [105/100]

File: app/services/notifiers/email_notification_service.rb [hunk 22]
```diff
@@ -41,4 +41,16 @@ def shortlisted_audit_certificate_reminder
...
     end
   end
+
+  def winners_notification
+    FormAnswer.winners.each do |form_answer|
+      email = if form_answer.promotion?
+        form_answer.document["nominee_email"]
+      else
+        form_answer.document["head_email"]
+      end
```
Comment (matthewford) [inline app/services/notifiers/email_notification_service.rb]: `end` at 51, 6 is not aligned with `if` at 47, 14

File: app/services/notifiers/email_notification_service.rb [hunk 23]
```diff
@@ -18,7 +18,7 @@ def run
   end
 
   # this will be removed after all methods are implemented
-  %w(reminder_to_submit ep_reminder_support_letters winners_notification winners_reminder_to_submit winners_press_release_comments_request unsuccessfull_notification all_unsuccessfull_feedback).each do |method|
+  %w(reminder_to_submit ep_reminder_support_letters winners_reminder_to_submit winners_press_release_comments_request unsuccessfull_notification all_unsuccessfull_feedback).each do |method|
```
Comment (matthewford) [inline app/services/notifiers/email_notification_service.rb]: Line is too long. [189/100]

File: app/services/notifiers/email_notification_service.rb [hunk 24]
```diff
@@ -41,4 +41,16 @@ def shortlisted_audit_certificate_reminder
       Notifiers::Shortlist::AuditCertificateRequest.new(form_answer).run
     end
   end
+
+  def winners_notification
+    FormAnswer.winners.each do |form_answer|
+      email = if form_answer.promotion?
+        form_answer.document["nominee_email"]
+      else
```
Comment (matthewford) [inline app/services/notifiers/email_notification_service.rb]: Align `else` with `if`.

File: app/services/notifiers/email_notification_service.rb [hunk 25]
```diff
@@ -41,4 +41,16 @@ def shortlisted_audit_certificate_reminder
       Notifiers::Shortlist::AuditCertificateRequest.new(form_answer).run
     end
   end
+
+  def winners_notification
+    FormAnswer.winners.each do |form_answer|
+      email = if form_answer.promotion?
+        form_answer.document["nominee_email"]
```
Comment (matthewford) [inline app/services/notifiers/email_notification_service.rb]: Use 2 (not -6) spaces for indentation.

File: app/services/notifiers/winners/buckingham_palace_invite.rb [hunk 26]
```diff
@@ -0,0 +1,10 @@
+class Notifiers::Winners::BuckinghamPalaceInvite
+  include Shoryuken::Worker
+
+  shoryuken_options queue: "default", auto_delete: true
+
+  def perform(sqs_msg, email)
```
Comment (matthewford) [inline app/services/notifiers/winners/buckingham_palace_invite.rb]: Unused method argument - `sqs_msg`. If it's necessary, use `_` or `_sqs_msg` as an argument name to indicate that it won't be used.

File: db/migrate/20150324104913_create_palace_attendees.rb [hunk 27]
```diff
@@ -0,0 +1,23 @@
+class CreatePalaceAttendees < ActiveRecord::Migration
+  def change
```
Comment (matthewford) [inline db/migrate/20150324104913_create_palace_attendees.rb]: Assignment Branch Condition size for change is too high. [17/15]

File: spec/mailers/users/buckingham_palace_invite_mailer_spec.rb [hunk 28]
```diff
@@ -0,0 +1,22 @@
...
+  describe "#notify" do
+    let(:mail) { Users::BuckinghamPalaceInviteMailer.invite(palace_invite.id) }
+
+    it "renders the headers" do
+      expect(mail.to).to eq([palace_invite.email])
+      expect(mail.from).to eq(["info@queensawards.org.uk"])
+    end
+
+    it "renders the body" do
+      expect(mail.body.encoded).to have_link("attendee details", href:  edit_palace_invite_url(id: palace_invite.token))
```
Comment (matthewford) [inline spec/mailers/users/buckingham_palace_invite_mailer_spec.rb]: Line is too long. [120/100]<br>Unnecessary spacing detected.

File: spec/services/notifiers/email_notification_service.rb [hunk 29]
```diff
@@ -55,4 +55,23 @@
       expect(current_notification.reload).to be_sent
     end
   end
+
+  context "winners_notifier" do
+    let!(:sent_notification) { create(:email_notification, trigger_at: Time.now - 1.day, sent: true, kind: "winners_notification") }
```
Comment (matthewford) [inline spec/services/notifiers/email_notification_service.rb]: Line is too long. [132/100]

File: spec/services/notifiers/email_notification_service.rb [hunk 30]
```diff
@@ -55,4 +55,23 @@
...
+    let!(:current_notification) { create(:email_notification, trigger_at: Time.now - 1.day, kind: "winners_notification") }
```
Comment (matthewford) [inline spec/services/notifiers/email_notification_service.rb]: Line is too long. [123/100]

File: spec/services/notifiers/email_notification_service.rb [hunk 31]
```diff
@@ -55,4 +55,23 @@
...
+    let!(:future_notification) { create(:email_notification, trigger_at: Time.now + 1.day, kind: "winners_notification") }
```
Comment (matthewford) [inline spec/services/notifiers/email_notification_service.rb]: Line is too long. [122/100]

File: spec/services/notifiers/email_notification_service.rb [hunk 32]
```diff
@@ -55,4 +55,23 @@
...
+    let(:user) { create(:user) }
+    let(:form_answer) { create(:form_answer, :submitted, document: { head_email: "head@email.com" }) }
```
Comment (matthewford) [inline spec/services/notifiers/email_notification_service.rb]: Line is too long. [102/100]

File: spec/services/notifiers/email_notification_service.rb [hunk 33]
```diff
@@ -55,4 +55,23 @@
...
+
+    it "triggers current notification" do
+      allow_any_instance_of(FormAnswer).to receive(:eligible?) { true }
+
+      expect(Notifiers::Winners::BuckinghamPalaceInvite).to receive(:perform_async).with("head@email.com")
```
Comment (matthewford) [inline spec/services/notifiers/email_notification_service.rb]: Line is too long. [106/100]

File: spec/services/notifiers/email_notification_service.rb [hunk 34]
```diff
@@ -2,9 +2,10 @@
 
 describe Notifiers::EmailNotificationService do
   context "shortlisted_audit_certificate_reminder" do
-    let!(:sent_notification) { create(:email_notification, trigger_at: Time.now - 1.day, sent: true, kind: "shortlisted_audit_certificate_reminder") }
-    let!(:current_notification) { create(:email_notification, trigger_at: Time.now - 1.day, kind: "shortlisted_audit_certificate_reminder") }
-    let!(:future_notification) { create(:email_notification, trigger_at: Time.now + 1.day, kind: "shortlisted_audit_certificate_reminder") }
+    let(:kind) { "shortlisted_audit_certificate_reminder" }
+    let!(:sent_notification) { create(:email_notification, trigger_at: Time.now - 1.day, sent: true, kind: kind) }
```
Comment (matthewford) [inline spec/services/notifiers/email_notification_service.rb]: Line is too long. [114/100]

File: spec/services/notifiers/email_notification_service.rb [hunk 35]
```diff
@@ -2,9 +2,10 @@
...
+    let!(:current_notification) { create(:email_notification, trigger_at: Time.now - 1.day, kind: kind) }
```
Comment (matthewford) [inline spec/services/notifiers/email_notification_service.rb]: Line is too long. [105/100]

File: spec/services/notifiers/email_notification_service.rb [hunk 36]
```diff
@@ -2,9 +2,10 @@
...
+    let!(:future_notification) { create(:email_notification, trigger_at: Time.now + 1.day, kind: kind) }
```
Comment (matthewford) [inline spec/services/notifiers/email_notification_service.rb]: Line is too long. [104/100]

File: spec/services/notifiers/email_notification_service.rb [hunk 37]
```diff
@@ -40,9 +51,10 @@
   end
 
   context "shortlisted_notifier" do
-    let!(:sent_notification) { create(:email_notification, trigger_at: Time.now - 1.day, sent: true, kind: "shortlisted_notifier") }
-    let!(:current_notification) { create(:email_notification, trigger_at: Time.now - 1.day, kind: "shortlisted_notifier") }
-    let!(:future_notification) { create(:email_notification, trigger_at: Time.now + 1.day, kind: "shortlisted_notifier") }
+    let(:kind) { "shortlisted_notifier" }
+    let!(:sent_notification) { create(:email_notification, trigger_at: Time.now - 1.day, sent: true, kind: kind) }
```
Comment (matthewford) [inline spec/services/notifiers/email_notification_service.rb]: Line is too long. [114/100]

File: spec/services/notifiers/email_notification_service.rb [hunk 38]
```diff
@@ -40,9 +51,10 @@
...
+    let!(:current_notification) { create(:email_notification, trigger_at: Time.now - 1.day, kind: kind) }
```
Comment (matthewford) [inline spec/services/notifiers/email_notification_service.rb]: Line is too long. [105/100]

File: spec/services/notifiers/email_notification_service.rb [hunk 39]
```diff
@@ -40,9 +51,10 @@
...
+    let!(:future_notification) { create(:email_notification, trigger_at: Time.now + 1.day, kind: kind) }
```
Comment (matthewford) [inline spec/services/notifiers/email_notification_service.rb]: Line is too long. [104/100]

File: spec/mailers/users/buckingham_palace_invite_mailer_spec.rb [hunk 40]
```diff
@@ -0,0 +1,23 @@
...
+    let(:mail) { Users::BuckinghamPalaceInviteMailer.invite(palace_invite.id) }
+
+    it "renders the headers" do
+      expect(mail.to).to eq([palace_invite.email])
+      expect(mail.from).to eq(["info@queensawards.org.uk"])
+    end
+
+    it "renders the body" do
+      expect(mail.body.encoded).to have_link("attendee details",
+                                             href:  edit_palace_invite_url(id: palace_invite.token))
```
Comment (matthewford) [inline spec/mailers/users/buckingham_palace_invite_mailer_spec.rb]: Unnecessary spacing detected.

File: app/services/notifiers/winners/buckingham_palace_invite.rb [hunk 41]
```diff
@@ -0,0 +1,10 @@
+class Notifiers::Winners::BuckinghamPalaceInvite
+  include Shoryuken::Worker
+
+  shoryuken_options queue: "default", auto_delete: true
```
Comment (RuslanHamidullin) [inline app/services/notifiers/winners/buckingham_palace_invite.rb]: Can you please correct Shoryuken Queue settings with:
```
shoryuken_options queue: "#{Rails.env}_default", auto_delete: true
```
As we are using Queue depends on ENV.
https://github.com/bitzesty/qae/blob/master/config/shoryuken.yml#L14
```
queues:
  - ["<%= Rails.env %>_mailers", 1]
  - ["<%= Rails.env %>_default", 2]
```

File: app/views/palace_invites/edit.html.slim [hunk 42]
```diff
@@ -0,0 +1,70 @@
...
+    article.group role="article"
+      .inner
+        p
+          ' On
+          = application_deadline "buckingham_palace_attendees_details"
+          '  there will be a Royal reception held at Buckingham Palace for organisations and individuals who have won Queen’s Awards. Winning organisations can send up to two attendees per award won.
+
+        br
+
+        = simple_form_for @invite_form, as: :palace_invite, url: palace_invite_path(id: @invite_form.invite.token), html: { class: 'qae-form', method: :patch } do |f|
```
Comment (RuslanHamidullin) [inline app/views/palace_invites/edit.html.slim]: Minor: form can be moved to partial

**Comment by RuslanHamidullin** [comment]
All good :+1: 
Need to correct queue name in https://github.com/bitzesty/qae/blob/palace-invites/app/services/notifiers/winners/buckingham_palace_invite.rb#L4

**Pull request merged**